### PR TITLE
More error tolerant web ui check

### LINKED ISSUE #1033: v7.6.2 Registry repeatedly fails to start, until it doesn't
Pelican Version:
v.7.6.2

Pelican Service:
- [ ] Client
- [ ] Plugin
- [X] Registry
- [ ] Director
- [ ] Origin
- [ ] Cache
- [ ] Other (please give the detail)

**Describe the bug**

>/tiger-osg-config$ kubectl -n osg logs osdf-registry-749c579558-9zmw8
>Starting Pelican...
>Running osdf with arguments: registry serve
>Logging.LogLocation is set to /var/log/pelican/osdf-registry.log. All logs are redirected to the log file.
>2024/04/02 19:42:08 goose: no migrations to run. current version: 20240212194951
>Error: Received bad status code in reply to server ping: 404. Expected 200,

On starting the v7.6.2 osdf-registry container, the container fails, but kubernetes re-starts it over and over. When this happened in the itb registry it only happened a few times and then settled down and served properly. It happened again when updating the prod registry, but while it settled down once, after accidentally restarting the container it continued to fail.

For the ITB osdf pelican regsitry, the last line about bad status codes did not appear on the successful run.

**To Reproduce**
Start/restart a v.7.6.2 registry container in kubernetes?

**Expected behavior**
The container should either always fail to start or always succeed, without needing to fail several times to serve properly.

**Desktop (please complete the following information):**
 - osdf-registry:v7.6.2 container

### PR DESCRIPTION
**Pull request by haoming29**
Should fix #1033 

The suspicion is that there's a racing condition at the registry start somehow, and the web engine is not "ready-ready" to serve requests, although it could be other issues. However, this PR will make the web ui check more robust in case of the error code returned

### CODE REVIEW

**Review by bbockelm — CHANGES_REQUESTED** [review]
I don't understand what this code is trying to do.

`WaitUntilWorking` tries a URL every 50ms until it succeeds or 10 seconds has passed.  It seems the code adds another set of loops to retry on top of retries (which appears strange?) ... no?

**Comment by haoming29** [comment]
> I don't understand what this code is trying to do.
> 
> `WaitUntilWorking` tries a URL every 50ms until it succeeds or 10 seconds has passed. It seems the code adds another set of loops to retry on top of retries (which appears strange?) ... no?

My guessing of William hitting that 404 error is that the gin server wasn't fully started before we hit the health check endpoint. Since 404 instead of 200 was returned, `WaitUntilWorking` returns immediately and we never get a chance to check it a bit later. I don't want to chance how `WaitUntilWorking` works because in most situations it makes sense to fail immediately if the status code is different from expected, but for the registry's case, it would probably worth waiting and checking again if the status code doesn't match

**Comment by bbockelm** [comment]
Gotcha.  I would rather add another option to WaitUntilWorking to make the wrong response code non-fatal.  Layers of retries are a bit scary.

**Comment by haoming29** [comment]
> Gotcha. I would rather add another option to WaitUntilWorking to make the wrong response code non-fatal. Layers of retries are a bit scary.

Sounds good. I'll refactor the code

**Comment by jhiemstrawisc** [comment]
@haoming29 is this ready for review or should I wait a bit longer?

**Comment by haoming29** [comment]
> @haoming29 is this ready for review or should I wait a bit longer?

Sorry I haven't get the chance to work on the refactoring. I will let you know when its' review for another round of review

**Comment by haoming29** [comment]
@jhiemstrawisc Should be good for the review. The Ubuntu test failure should be a random one and re-running it will likely fix the issue

**Review by bbockelm — APPROVED** [review]
LGTM

File: server_utils/server_utils.go [hunk 1]
```diff
@@ -75,24 +79,55 @@ func WaitUntilWorking(ctx context.Context, method, reqUrl, server string, expect
 				}
 				bytes, err := io.ReadAll(resp.Body)
 				if err != nil {
-					// We didn't get the expected status
-					return errors.Errorf("Received bad status code in reply to server ping at %s: %d. Expected %d. Can't read response body with error %v.", reqUrl, resp.StatusCode, expectedStatus, err)
+					statusError = errors.Errorf("Received bad status code in reply to server ping at %s: %d. Expected %d. Can't read response body with error %v.", reqUrl, resp.StatusCode, expectedStatus, err)
+					if statusMismatch {
+						if !statusErrLogged {
+							log.Info(statusError, "Will retry until timeedout")
```
Comment (jhiemstrawisc) [inline server_utils/server_utils.go:85]: timeedout --> timeout

File: server_utils/server_utils.go [hunk 2]
```diff
@@ -41,13 +41,17 @@ import (
 
 // Wait until given `reqUrl` returns a HTTP 200.
 // Logging messages emitted will refer to `server` (e.g., origin, cache, director)
-func WaitUntilWorking(ctx context.Context, method, reqUrl, server string, expectedStatus int) error {
+// Pass true to statusMismatch to allow a mismatch of expected status code and what's returned not fail immediatelys
```
Comment (jhiemstrawisc) [inline server_utils/server_utils.go:44]: immediatelys --> immediately

File: server_utils/server_utils.go [hunk 3]
```diff
@@ -75,24 +79,55 @@ func WaitUntilWorking(ctx context.Context, method, reqUrl, server string, expect
...
+						return statusError
+					}
 				} else {
 					if len(bytes) != 0 {
-						// We didn't get the expected status
-						return errors.Errorf("Received bad status code in reply to server ping at %s: %d. Expected %d. Response body: %s", reqUrl, resp.StatusCode, expectedStatus, string(bytes))
+						statusError = errors.Errorf("Received bad status code in reply to server ping at %s: %d. Expected %d. Response body: %s", reqUrl, resp.StatusCode, expectedStatus, string(bytes))
+						if statusMismatch {
+							if !statusErrLogged {
+								log.Info(statusError, "Will retry until timeedout")
```
Comment (jhiemstrawisc) [inline server_utils/server_utils.go:97]: Same here

File: server_utils/server_utils.go [hunk 4]
```diff
@@ -75,24 +79,55 @@ func WaitUntilWorking(ctx context.Context, method, reqUrl, server string, expect
...
+						} else {
+							// We didn't get the expected status
+							return statusError
+						}
 					} else {
-						return errors.Errorf("Received bad status code in reply to server ping at %s: %d. Expected %d. Response body is empty.", reqUrl, resp.StatusCode, expectedStatus)
+						statusError = errors.Errorf("Received bad status code in reply to server ping at %s: %d. Expected %d. Response body is empty.", reqUrl, resp.StatusCode, expectedStatus)
+						if statusMismatch {
+							if !statusErrLogged {
+								log.Info(statusError, "Will retry until timeedout")
```
Comment (jhiemstrawisc) [inline server_utils/server_utils.go:108]: Here too

**Review by jhiemstrawisc — CHANGES_REQUESTED** [review]
A few small typos, otherwise LGTM

**Comment by haoming29** [comment]
> A few small typos, otherwise LGTM

Thanks for catching them!

**Review by jhiemstrawisc — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**